elf2cfetbl_version.h

```diff
@@ -0,0 +1,64 @@
+/*
+**  GSC-18128-1, "Core Flight Executive Version 6.7"
+**
+**  Copyright (c) 2006-2019 United States Government as represented by
+**  the Administrator of the National Aeronautics and Space Administration.
+**  All Rights Reserved.
+**
+**  Licensed under the Apache License, Version 2.0 (the "License");
+**  you may not use this file except in compliance with the License.
+**  You may obtain a copy of the License at
+**
+**    http://www.apache.org/licenses/LICENSE-2.0
+**
+**  Unless required by applicable law or agreed to in writing, software
+**  distributed under the License is distributed on an "AS IS" BASIS,
+**  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+**  See the License for the specific language governing permissions and
+**  limitations under the License.
+*/
+
+/*
+ *  File: elf2cfetbl_version.h
+ *
+ *  Purpose:
+ *  Provide version identifiers for the ELF to cFE Table Converter.
+ *  See cfe documentation for version and build number and description
+ */
+
+#ifndef ELF2CFETBL_VERSION_H
+#define ELF2CFETBL_VERSION_H
+
+/*
+ * Development build identifiers
+ */
+#define ELF2CFETBL_CODENAME "Bootes"
+#define ELF2CFETBL_BASELINE "v3.1.0"
+#define ELF2CFETBL_BUILDNUMBER 34
+/*
+ * Version Macro Definitions
+ * ONLY APPLY if CFS_RELEASE_BUILD == TRUE
+ */
+#define ELF2CFETBL_MAJOR_VERSION 3
+#define ELF2CFETBL_MINOR_VERSION 1
+#define ELF2CFETBL_REVISION 5
+#define ELF2CFETBL_MISSION_REV 0
```

Comment:
I know I previously said I'd drop this issue, but this still seems way over complicated.

As the "baseline" string is there - can't we just display the baseline string and remove the individual numeric major/minor numbers.  It once again seems like repeating the same info but unclear why we have both or why they'd be different.

Reply:
That's what I'd like to do but there was apprehension about removing the numeric version numbers so I'm leaving them as a compromise.

Reply:
I dunno - seems like we are just way over complicating this - it is such a simple idea, put a string into the code which serves as it's self-identification and display said string when requested.  

Problem solved. Occam's Razor - No more complicated than it needs to be.

There is one justified bit of added complexity - and that's to identify a development/non-release build and make it also self report as such - which I thought was accomplished by adding "+dev##" suffix/number - but that should also be just one macro.

I don't want to get into another back-and-forth argument over this, but after mulling on it for a few days -- please lets just keep it as simple as possible.  You can do anything you want with just a simple string.  That's my vote.

Reply:
I like simple. Maybe - "Development: *tag*-dev##, last official release vX.Y.Z", and "Official release: vX.Y.Z" where everything except ##, X, Y, and Z are just part of the string?

Reply:
Still sounds complicated - I thought the runtime self-reporting requirement was strictly to identify the current version of the code that's being executed (be it development or official).  

Is there also a requirement that the code also separately self-identify the previous official release, if it is not an official release itself?  (honest question - just want to be sure we aren't adding complexity that isn't backed by a requirement).

If that is a requirement - then fine - but if not this is even simpler:

```
#define ELF2CFETBL_VERSION_BASELINE "v3.1.0"
#define ELF2CFETBL_VERSION_SUFFIX     "+dev42"
```

Where the SUFFIX could be a string like "(official)" when it is actually an official release tag with no dev commits on top of it, instead of a dev##.

The main thing is that these strings in the header(s) are simply the RESULT of whatever arbitrarily complex logic is used to determine the "version number" it reports (be it semver, calver, or whatever other legal/extra rules we need to adhere to).    We should just try to keep that complexity outside the source code/headers, that's all.  (and likewise it should be the minimum of strings required to meet the applicable self-reporting requirement).

Reply:
For a development version the software shall report on execution:
  - that it is a development version
  - the last official release the development version is based on
  - whatever it takes to **uniquely** identify the development version.

For an official release, the software shall report on execution:
  - that it is an official release
  - the official release version.

Those are the requirements I suggest we impose on this implementation to easily identify code in the wild.

Reply:
OK - then this information above is what I'd suggest putting in the code but just as _strings_ - not integer numbers.  The strings can be in the form "major.minor.rev" but this affords the opportunity for other nomenclatures/schemes without changing the general paradigm, so we can (theoretically) use the same update script/tool on every repo.

I do think that one can achieve objectives of identifiable/unique development version as well as differentiating a dev version from an official version with just one string  - the "suffix" I suggested above does this (I think?), could be something else, just trying to keep it simple and avoid redundant info in code.

Note there are plenty of other examples where we had e.g. "v6.5.0a" or OSAL "4.2.1a" which was typically some minor documentation or configuration patch after the original release was already tagged.  Having a string, rather than strictly three integers also allows this possibility, as well has having the string exactly match the github tag of said release, which IMO is another nice-to-have.

Reply:
@skliper I suggest adding a version information requirement in the source code as well

Reply:
I like moving away from the integers. What functionality are we losing by switching those to strings?

Reply:
IIRC the objection to moving away from integers is that it's not possible to do a "Version check" macro - i.e. some dependent code could do e.g. `#if (VERSION_MAJOR > 3) || ((VERSION_MAJOR >= 3) && (VERSION_MINOR >= 1) )` if it knows that a particular API/feature it wants to use was added in version 3.1 of said tool/library.

But we don't typically do this anywhere, nor would I recommend doing that - it can get ugly.  I think the constellation name branches can basically do the same thing without the ifdefs.

Reply:
Let's keep them for now and remove them on 7.0 since we don't know if there are people using these defines out in the wild